flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/HsMemoryDataManager.java

```diff
@@ -0,0 +1,673 @@
+                            consumeBuffers.add(bufferContext);
+                            bufferIndexToContexts.put(
+                                    bufferContext.bufferIndexAndChannel.getBufferIndex(),
+                                    bufferContext);
+                            if (consumeBuffers.isEmpty()) {
+                                needNotify = true;
+                            }
+                        }
+                        if (needNotify) {
+                            // TODO notify data available.
+                        }
+                    });
+        }
+
+        @GuardedBy("subpartitionLock")
+        private boolean isBufferSatisfyStatus(
+                BufferContext bufferContext, SpillStatus spillStatus, ConsumeStatus consumeStatus) {
+            // released buffer is not needed.
+            if (bufferContext.isReleased) {
+                return false;
+            }
+            boolean isNeeded = true;
+            switch (spillStatus) {
+                case NOT_SPILL:
+                    isNeeded = !bufferContext.isSpill;
+                    break;
+                case SPILL:
+                    isNeeded = bufferContext.isSpill;
+                    break;
+            }
+            switch (consumeStatus) {
+                case NOT_CONSUMED:
+                    isNeeded &= !bufferContext.isConsumed;
+                    break;
+                case CONSUMED:
+                    isNeeded &= bufferContext.isConsumed;
+                    break;
+            }
+            return isNeeded;
+        }
+    }
+
+    /**
+     * This class maintains the buffer information and its status for hybrid shuffle mode.
+     *
+     * <p>Note that: This class itself is not thread-safe, guarded by subpartition lock.
+     */
+    private static class BufferContext {
+        private final Buffer buffer;
+
+        private final BufferIndexAndChannel bufferIndexAndChannel;
+
+        private boolean isReleased;
+
+        private boolean isSpill;
+
+        private boolean isConsumed;
+
+        private CompletableFuture<Void> spilledFuture;
+
+        public BufferContext(Buffer buffer, int bufferIndex, int subpartitionId) {
+            this.bufferIndexAndChannel = new BufferIndexAndChannel(bufferIndex, subpartitionId);
+            this.buffer = buffer;
+        }
+
+        public void onBufferReleased() {
+            isReleased = true;
+            buffer.recycleBuffer();
+        }
+
+        public void onBufferSpill(CompletableFuture<Void> spilledFuture) {
+            // retainBuffer and mark buffer is spill status.
+            buffer.retainBuffer();
+            isSpill = true;
+            checkState(this.spilledFuture == null, "Already set spill future for buffer context.");
+            this.spilledFuture = spilledFuture;
+        }
+
+        public void onBufferSpilled() {
+            // decrease ref count when spilling is finished.
+            buffer.recycleBuffer();
+        }
```

Comment:
Why do we need this method? Can this be scheduled in `onBufferSpill`?

Reply:
yes, scheduled it to spillFuture